extensions/amp-lightbox-gallery/1.0/storybook/Basic.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../../src/preact';
+// import {LightboxGallery} from '../lightbox-gallery';
+import {Lightbox} from './../../../amp-lightbox/1.0/lightbox';
+import {withA11y} from '@storybook/addon-a11y';
+import {withKnobs} from '@storybook/addon-knobs';
+
+export default {
+  title: 'LightboxGallery',
+  // component: LightboxGallery,
+  decorators: [withKnobs, withA11y],
+};
+
+const LightboxGalleryContext = Preact.createContext({});
+
+/**
+ * @param {!Object} props
+ * @return {*}
+ */
+function WithLightbox({children}) {
+  const {open, pushChild} = Preact.useContext(LightboxGalleryContext);
+  Preact.useLayoutEffect(() => {
+    pushChild(children);
+  }, []);
+  return <div onClick={() => open()}>{children}</div>;
+}
+
+/**
+ * @param {!Object} props
+ * @return {*}
+ */
+function LightboxGallery({children}) {
+  const lightboxRef = Preact.useRef(null);
+  const lightboxElements = Preact.useRef([]);
+  const pushChild = (el) => lightboxElements.current.push(el);
+  const context = {pushChild, open: () => lightboxRef.current.open()};
+  return (
+    <>
+      <Lightbox ref={lightboxRef}>
+        <div onClick={() => lightboxRef.current.close()}>close lightbox</div>
+        <div>{lightboxElements.current}</div>
+      </Lightbox>
+      <LightboxGalleryContext.Provider value={context}>
+        {Preact.toChildArray(children).map((child) =>
+          child.props.lightbox ? <WithLightbox>{child}</WithLightbox> : child
+        )}
+      </LightboxGalleryContext.Provider>
+    </>
+  );
+}
+
+export const _default = () => {
+  return (
+    <LightboxGallery>
```

Comment:
Ok, I think I understood your process here. And it makes sense to me. One nuance - it cannot traverse nested children. But we could enable it with making `WithLightbox` public. E.g. the user code could look like this:

Somewhere once:
```
<LightboxGallery>  // could also call it `(Auto)LightboxProvider`
  {children}
</LightboxGallery>
```

And elsewhere:

```
function SomeComponent() {
  return (
    <div>
      <WithLightbox>
         <img ..../>
      </WithLightbox>
    </div>
  );
}
```

This might be pretty clean from Preact point of view. Some of our components could even call `WithLightbox` internally.

In AMP we could enable this somehow too:
1. The `LightboxProvider` attachment is very easy for us now - all it needs to do is to propagate context. We do this today with `InlineGallery`.
2. The `WithLightbox` would be a bit harder. But I think we can figure it out with "detached" mode. Or maybe we'd just always require related components to call `WithLightbox` internally for now.

The important question here: do we currently clone DOM elements to put them inside the lightbox? Or do we reparent them? Or just position? This is important, because if we clone/reposition, then in React that'd probably be taken care of by portalling, which we haven't worked with yet here, but could open some doors for us.

Reply:
We can get nested children if we are willing to clone elements, but this is costly and somewhat hacky to drill it down like this:
```jsx
<LightboxGalleryContext.Provider value={context}>
  <WithDeepLightbox>{children}</WithDeepLightbox>
</LightboxGalleryContext.Provider>
```

```js
function WithDeepLightbox({children}) {
  return toChildArray(children).map((child) =>
    child.props?.lightbox ? (
      <WithLightbox>{child}</WithLightbox>
    ) : child.props?.children ? (
      Preact.cloneElement(child, {
        children: <WithDeepLightbox>{child.props.children}</WithDeepLightbox>,
      })
    ) : (
      child
    )
  );
}
```

To your point about portals - 0.1 clones and cleans lightboxable DOM elements:
https://github.com/ampproject/amphtml/blob/ea2e7f696feca04e4fc0be41f4dc60c9a240904d/extensions/amp-lightbox-gallery/0.1/amp-lightbox-gallery.js#L269-L290 Can you describe more how portals helps this use case? I imagine we would want lightboxed elements both to be rendered where they are in the DOM as well as in the LightboxGallery when it is opened. From what I've read about portals, it doesn't seem like you can have it both ways?